### PR TITLE
Translation function for qubit operators

### PR DESCRIPTION
**Pull request by AlexandreF-1qbit**
Translation function to convert qubit operators.

e.g.
```python
tangelo_op = translate_operator(qiskit_op, source="qiskit", target="tangelo")
```

### CODE REVIEW

File: tangelo/toolboxes/operators/translation.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+from tangelo.toolboxes.operators import count_qubits, QubitOperator
+from tangelo.linq.helpers import pauli_of_to_string, pauli_string_to_of
+
+
+SUPPORTED_SOURCES = {"tangelo", "qiskit"}
+SUPPORTED_TARGETS = {"tangelo", "qiskit"}
+
+
+def translate_operator(qubit_operator, source, target, n_qubits=None):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/translation.py:26]: I suggest "tangelo" for the default value of `source`.

File: tangelo/toolboxes/operators/translation.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    supported.
+
+    Args:
+        qubit_operator (source format): Self-explanatory.
+        source (string): Identifier for the source format.
+        target (string): Identifier for the target format.
+        n_qubits (int): Number of qubits relevant to the operator.
+
+    Returns:
+        (taget format): Translated qubit operator.
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/translation.py:38]: typo

File: tangelo/toolboxes/operators/translation.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+from tangelo.linq.helpers import pauli_of_to_string, pauli_string_to_of
+
+
+SUPPORTED_SOURCES = {"tangelo", "qiskit"}
+SUPPORTED_TARGETS = {"tangelo", "qiskit"}
+
+
+def translate_operator(qubit_operator, source, target, n_qubits=None):
+    """Function to convert a qubit operator defnied within the "source" format
+    to another format. Only the trnaslation from and to tangelo are currently
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/translation.py:28]: typo

File: tangelo/toolboxes/operators/translation.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    for term_tuple, coeff in qubit_operator.terms.items():
+        term_string = pauli_of_to_string(term_tuple, n_qubits)
+
+        # Reverse the string becasue of qiskit convention.
+        term_list  += [(term_string[::-1], coeff)]
+
+    return PauliSumOp.from_list(term_list)
+
+
+def qiskit_to_tangelo(qubit_operator):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/translation.py:85]: Let's include automated tests for simple operators, but also try large ones to see if performance is ok. As long as the tests are a few seconds, they are fine to include in automated tests.

File: tangelo/toolboxes/operators/translation.py (hunk 1 above)
Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/operators/translation.py:26]: A way to handle the 3 (4) cases (which gives you 3 different scenario in your tests). Assume source=s, target=t (s, t != "tangelo").

- Case 0: if source and target are identical, do nothing, return the input operator immediately
- Case 1: tangelo -> t
- Case 2: s -> tangelo
- Case 3: s -> t

===============

```
op = input_op # was provided as parameter to function by user
if s != "tangelo":
   op = <code to convert from source s to tangelo format>
if t != "tangelo":
  op = <code to convert from tangelo to target t format>
return op
```

File: tangelo/toolboxes/operators/translation.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    terms_dict = dict()
+    for pauli_word in qubit_operator:
+        # Inversion of the string because of qiskit ordering.
+        term_string = pauli_word.to_pauli_op().primitive.to_label()[::-1]
+        term_tuple = pauli_string_to_of(term_string)
+        terms_dict[tuple(term_tuple)] = pauli_word.coeff
+
+    # Create and copy the information into a new QubitOperator.
+    tangelo_op = QubitOperator()
+    tangelo_op.terms = terms_dict.copy()
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/translation.py:106]: why are we making a copy ? Can't we just assign it ?

Comment (AlexandreF-1qbit) [reply tangelo/toolboxes/operators/translation.py]: I remember having problems in the past with dictionaries. I removed it but if the implementation does not work, I will check this line first.

File: tangelo/toolboxes/operators/translation.py (hunk 1 above)
Comment (AlexandreF-1qbit) [reply tangelo/toolboxes/operators/translation.py:30]: Tell me if I understood well with the changes I made. I think it is cleaner, thanks for the suggestion.

File: tangelo/toolboxes/operators/translation.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+TO_TANGELO = {
+    "qiskit": qiskit_to_tangelo
+}
+
+
+def translate_operator(qubit_operator, source, target, n_qubits=None):
+    """Function to convert a qubit operator defined within the "source" format
+    to another format. Only the trnaslation from and to tangelo are currently
```
Comment (JamesB-1qbit) [inline tangelo/toolboxes/operators/translation.py:32]: trnaslation -> translation

**Review by JamesB-1qbit — APPROVED** [review]

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
I think I'd put the new functionalities under linq.translator, to make it easy for users.

Linq's scope is connecting our platform to others, so this makes sense.

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you Alexandre !

**Pull request merged**